### DISCUSSION TITLE
Proxmox VE LXC Filesystem Trim

### DISCUSSION THREAD

**Discussion by tteck** [tteck/Proxmox#2505] [Announcements]
## Proxmox VE LXC Filesystem Trim

With traditional hard disk drives (HDDs or spinner disks), there isn't a significant performance penalty for leaving empty blocks or unused space on the disk. This is because HDDs use spinning magnetic platters and mechanical read/write heads to access data. Therefore, the physical location of data on the disk doesn't impact performance significantly, and there's no need to actively manage unused space to maintain performance.

However, solid-state drives (SSDs) operate differently. SSDs use NAND flash memory to store data, and they have a limited number of write cycles per cell. As data is written, erased, and rewritten to the SSD, performance can degrade over time due to the way NAND flash memory works. Specifically, write operations are faster on empty blocks than on blocks that have been previously written and need to be erased before new data can be written.

`fstrim` is a command used with SSDs to inform the drive which blocks of data are no longer considered in use by the filesystem so that the drive can internally erase them and prepare them for new data. This helps maintain consistent performance on SSDs over time and prevents degradation due to inefficient write operations.

Similarly, thin-provisioned storage technologies like LVM-Thin, Ceph, and qcow2 allocate storage space dynamically as needed. Without proper management, unused blocks within these storage systems can continue to occupy space, causing storage usage to grow unnecessarily.

Therefore, it's essential to periodically trim SSDs and manage unused space in thin-provisioned storage systems to optimize performance and prevent unnecessary consumption of storage resources.

VMs are configured to run `fstrim` automatically on a weekly basis, for LXC containers, it's necessary to perform `fstrim` manually or set up a separate automated process.

So, I created a script to run `fstrim` on all LXC containers (that are not selected to be skipped). This is designed to work with _SSDs_ on **ext4 filesystems only**.

So, lets test it 
```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/misc/fstrim.sh)"
```

**Comment by jcastro** [comment]
Ran it on my Prox 8 for docker, npm and adguard LXC. All went ok! thanks

**Comment by joyshmitz** [comment]
Tried it on my home lab ProxMox 8, 7 containers, worked great.

Thanks for the help.

**Comment by 0xk1f0** [comment]
Trimmed all of my 8 LXC-Containers on my Proxmox 8 instance without issues.

Neat script, thanks!

**Comment by towerhand** [comment]
There were no errors on my end running Proxmox 8.1.4 with 24 LXC containers. Thanks for the script.

I think this is a good candidate for an unintended cron script since you mentioned VMs are configured to run fstrim automatically weekly.

**Reply by tteck** [reply]
Something like:
```
30 0 * * 0 sh -c 'pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do template=$(pct config $ct | grep -q "template:" && echo "true" || echo "false"); if [ "$template" = "false" ]; then pct fstrim $ct; fi; done >> /var/log/fstrim-cron.log 2>/dev/null'
```
(template check)

or
```
30 0 * * 0 sh -c 'pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do pct fstrim $ct; done >> /var/log/fstrim-cron.log 2>/dev/null'
```
(no template check)

set to run at 12:30 AM every Sunday

Logs can be viewed by executing the command `cat /var/log/fstrim-cron.log`

**Reply by towerhand** [reply]
Awesome. Thanks.

**Reply by eikaramba** [reply]
strange, when i execute that command(with template check) in the shell it works, however in the crontab i get a `sh: 1: pct: not found` error. any idea why?

**Reply by divemasterjm** [reply]
me too

**Reply by tteck** [reply]
Let's start with a plain cron job, nothing fancy.
```
30 0 * * 0 pct list | awk '/^[0-9]/ {print $1}' | while read ct; do pct fstrim ${ct}; done
```

**Reply by tteck** [reply]
For a test:
```
30 0 * * 0 pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do template=$(pct config $ct | grep -q "template:" && echo "true" || echo "false"); if [ "$template" = "false" ]; then pct fstrim $ct; fi; done >> /var/log/fstrim-cron.log 2>/dev/null
```

**Reply by eikaramba** [reply]
i tested the first plain cron job and got this result:
`/bin/sh: 1: pct: not found`

the same for the second one:
`/bin/sh: 1: pct: not found`

**Reply by tteck** [reply]
Where are you creating the crontab? Proxmox host?

**Reply by eikaramba** [reply]
yes exactly - very weird. i haven't found any clue why i get this error in the crontab but the command executed in shell works flawessly it seems

**Reply by Xpl0itU** [reply]
The error happens because cron doesn't have a PATH set, so you have to include the full path of the pct binary, this crontab entry should work
```
30 0 * * 0 /usr/sbin/pct list | awk '/^[0-9]/ {print $1}' | while read ct; do /usr/sbin/pct fstrim ${ct}; done
```

**Comment by onethree7** [comment]
you could add that your script can take -f option (like 'force trim all' , if this is chosen whiplash gets bypassed).
then you can offer users a dialog in the normal first run of the script via whiplash "would you like to add this script to crontab" the script could write itself into crontab with the -f option then. (or place a manual method like in your previous comment, and dont bother -f either) :D I dont think the whiplash dialog is neccessary at all here generally. Who would want to skip an LXC 

I think thats only natural and most will choose this workflow and cron-setup.

Thanks for the good work.

**Reply by tteck** [reply]
🤣 It's called "whiptail"

**Reply by onethree7** [reply]
sh1t really :D Sorry i only used dialog on bsd before

whiplash sounds cooler though

**Comment by Mellowlynx** [comment]
Thanks, Ran without problems here.

**Comment by brusarp** [comment]
Very helpful! Thanks... However, I think it is better to correct some terms... In the first window of the script it says "The selected containers will be subjected to..." In the second one it is asked to "select those to exclude..." A sort of linguistic short circuit and cause of possible misunderstandings in my opinion...

**Reply by tteck** [reply]
Corrected, first window now, "The LXC containers will undergo the fstrim command. Proceed?".

**Comment by tteck** [comment]
The script now displays the data percentage used before and after trimming.

**Comment by Calimerorulez** [comment]
On some of my LXCs, percentage before and after is shown as empty. Only a percentage sign is displayed.

**Reply by tteck** [reply]
What's the difference between the LXCs?

**Reply by Calimerorulez** [reply]
Good question.

One of them that isn't showing is the Node-Red LXC, and another one is the Homebridge LXC. The Plex LXC doesn't show the percentage either.

**Reply by tteck** [reply]
I have no idea why those don't show the info.
All of mine show the info, including my node-red LXC.

**Reply by MickLesk** [reply]
ive 2 examples @tteck:

1st: You have an LXC running on SSD but its stopped / shutdown | Percent shown correctly but hostname not
![image](https://github.com/tteck/Proxmox/assets/47820557/23376b5c-3455-42cb-8275-30bb536ab740)

2nd: You have an LXC running on an HDD, then the status (Running, Stopped...) does not matter. 
![image](https://github.com/tteck/Proxmox/assets/47820557/c7c93c16-c707-448d-86e1-f21a2cf948b5)

**Reply by tteck** [reply]
Both examples demonstrate that the script handled everything as expected.

**Reply by MickLesk** [reply]
@tteck  
Isn't it better to make the error handling a little clearer?
Say like in your "LXC update script". Start 101... (then the hostname is also displayed) -> solution for 1st issue

And for the 2nd as error message before "Data before trim %" come, HDD Detected, Skip...?

**Reply by tteck** [reply]
Unlike the update script, containers do not need to be running to be trimmed. (maybe show CTID instead of hostname)
I'll look into device detection, but the script allows for containers to be skipped.

I think both examples explain the situation clearly.

**Reply by tteck** [reply]
The script now displays the container ID instead of the hostname. https://github.com/tteck/Proxmox/commit/b082f8a08e3b4e597bf069306194f597709c469c

By the way, neither of your examples demonstrates an error, they purely provide information.

**Reply by Calimerorulez** [reply]
My 'problematic' LXCs do run of a SSD too, not from a HDD, fwiw.

**Reply by Calimerorulez** [reply]
The ones that don't show the percentages are running on the same Samsung SSD. Maybe that's why.

**Reply by tteck** [reply]
I bet the Samsung SSD is set as type Directory instead of LVM-Thin.

**Reply by Calimerorulez** [reply]
You are right! :)

**Reply by tteck** [reply]
If the contents of that Samsung SSD aren't critical to you, consider repurposing it as a thinpool for efficient thin provisioning. This can be easily achieved through the GUI.

**Comment by divemasterjm** [comment]
for me it worked

**Comment by divemasterjm** [comment]
this one doenst work for me.
Cron ***@***.***> pct list | awk "/^[0-9]/ {print \$1}" | while read ct; do
template=$(pct config $ct | grep -q "template:" && echo "true" || echo
"false"); if [ "$template" = "false" ]; then pct fstrim $ct; fi; done >>
/var/log/fstrim-cron.log 2>/dev/null

/bin/sh: 1: pct: not found

**Comment by AJolly** [comment]
So what happens if you're running ZFS?

**Comment by nirav-mj** [comment]
Sorry I'm new to this, but am I supposed to run the script in LXC container which is running Ubuntu or do I need run this on the Proxmox host? Thanks!

**Reply by tteck** [reply]
Run in the Proxmox shell

**Reply by nirav-mj** [reply]
Thank you!! so the script will auto detect my LVM-THIN SSD disk and start the trim process?

**Reply by nirav-mj** [reply]
I just ran it, and it seem to work. Is there way to see the details how much was trimmed? the screen went by quickly

**Reply by tteck** [reply]
> Is there way to see the details

It doesn't store logs.
> the screen went by quickly

Not so quick that you couldn't read the before/after stats.

**Reply by nirav-mj** [reply]
lol, I was using Guacamole SSH connection and it was quick, anyway I ran it again using putty and I could see it. Thank you very much for all your time and effort you deserve a coffee.

**Comment by aroundmyroom** [comment]
could the bytes saved in human readable form like 10GB / 100MB? one LXC went from 99,5% to 9.3% ... ;)

**Reply by tteck** [reply]
![Screenshot 2024-04-11 7 24 02 PM](https://github.com/tteck/Proxmox/assets/21342670/1c95888c-b289-43ae-a25e-c1dfd43d5706)

**Comment by bazb** [comment]
Please add a cron version to the .io page

**Reply by neyzm** [reply]
+1, would be super useful.

**Comment by NeoPlayer13** [comment]
Or a way to add the containers to skip as a shell parameter.

**Comment by AT-StephenDetomasi** [comment]
Is this script actually needed on ZFS? As long as you trim the host disks, running additional trims for LXCs should not be necessary as ZFS is a block-level filesystem. Happy to be corrected though.

**Reply by tteck** [reply]
Read the instructions.

**Comment by pylchung** [comment]
When I run the script I get a message that root file system is not ext4. My boot SSD is ZFS, but my SSD for VMs/LXC are LVM-Thin. Can I not use this script?

**Reply by tteck** [reply]
You cannot, sorry.

**Reply by pylchung** [reply]
Thanks for confirming. I managed to run trim on each LXC manually using `pct fstrim <ctid>`.
